working/0508 - strict-inference/strict-inference.md

```diff
@@ -0,0 +1,350 @@
+
+> T fold <T>(T initialValue, T combine(T previousValue, E element))
+
+```dart
+void main() {
+  var a = [1, 2, 3].fold(true, (s, x) => s + x);
+}
+```
+
+There are no compile-time, static analysis errors in this code, but it is
+guaranteed to produce a failure at runtime, when `true + 1` is executed. The
+issue here is similar to the previous one: the developer likely thinks that the
+type of `s` (`T`) will be inferred from the type of `initialValue`, and that
+static analysis would report any issue with that type. But inference doesn't
+flow between parameters like that.  Instead, while trying to infer `T`, there
+is not enough information from downwards inference (the type of `a`), and there
+is not enough information from upwards inference (the type of `(s, x) => s +
+x`). Inference gives up on the type of the function literal, and gives it
+`dynamic Function(dynamic, int)`. Inference combines that type with `bool`, the
+type of `true`, and settles on `dynamic` for `T`.
+
+The issue would be revealed with either an explicit type on `fold`, an explicit
+type for `a`, which would help to infer the type of `s`, or an explicit type
+for `s`:
+
+```dart
+void main() {
+  // Each of these produce an existing error:
+  // "The operator '+' isn't defined for the class 'bool'."
+  var b = [1, 2, 3].fold<bool>(true, (s, x) => s + x);
+  bool c = [1, 2, 3].fold(true, (s, x) => s + x);
+
+  // This produces an existing error:
+	// "Couldn't infer type parameter 'T'. Tried to infer 'Object' for 'T' which
+  // doesn't work: Parameter 'combine' declared as 'T Function(T, int)' but
+  // argument is 'int Function(int, int)'. The type 'Object' was inferred from:
+  // Parameter 'initialValue' declared as 'T' but argument is 'bool'. Consider
+  // passing explicit type argument(s) to the generic."
+  var d = [1, 2, 3].fold(true, (int s, x) => s + x);
+}
+```
+
+In strict inference mode, the inference failure on `(s, x) => s + x` will be
+reported, enouraging the developer to add a type to `s`, `a`, or `fold`,
+revealing their misunderstaing of the types.
+
+[`Iterable.fold`]: https://api.dartlang.org/stable/dart-core/Iterable/fold.html
+
+## Conditions for strict inference failure
+
+This is an exhaustive list of conditions that result in an inference failure,
+under the strict inference mode. Examples are given for each condition, as
+well as examples that highlight code without any inference failures. The
+location of the failure is listed for each condition.
+
+### Uninitialized variable
+
+A variable or field declared without a type (via `var` or `final`) and without
+an initializer is considered an inference failure.
+
+```dart
+void main() {
+  var x;        // Inference failure
+  var y = 7;    // OK
+}
+
+class C {
+  final f;      // Inference failure
+  final g = 7;  // OK
+
+  C(this.f);
+
+  static var s; // Inference failure
+}
+```
+
+The failure is reported on the identifier of the variable or field.
```

Comment:
We have generally taken the approach that language documents specify what is an isn't valid (in this case, what is or isn't an inference failure), but leave details such as where the error is reported to the tools to decide. I'd prefer that we drop lines like this one from this specification.

Reply:
Fine by me! Done.